hnn_core/cell.py

```diff
@@ -231,6 +234,41 @@ def insert_dipole(self, yscale):
             dpp.ztan = y_diff[-1]
         self.dipole = h.Vector().record(self.dpl_ref)
 
+    def create_tonic_feed(self, amplitude, t0, T, loc=0.5):
+        """Create tonic feed at the soma.
+
+        Parameters
+        ----------
+        amplitude : float
+            The amplitude of the input.
+        t0 : float
+            The start time of tonic input (in ms).
+        T : float
+            The end time of tonic input (in ms).
+        loc : float (0 to 1)
+            The location of the input in the section.
+        """
+        duration = T - t0
+
+        if T < 0.:
+            raise ValueError('End time of tonic input cannot be negative')
+
+        if duration < 0.:
+            raise ValueError('Duration of tonic input cannot be negative')
+
```

Comment:
These are the sanity checks needed. I'd much rather have them already in `Network`! See my other inline comments. How about something like:

```python
# in cell.py
def _check_tonic_feed_values(amplitude, t0, T, loc=0.5):
    if T < 0.:
    ....

# in network.py
from .cell import _check_tonic_feed_values
tonic_params= {
                      'amplitude': self.params[f'Itonic_A_{cell_type}_soma'],
                      't0': self.params[f'Itonic_t0_{cell_type}_soma'],
                      'T': self.params[f'Itonic_T_{cell_type}_soma']
                  }
_check_tonic_feed_values(**tonic_params)
```

Reply:
Any reason for storing the function in `cell.py` instead of `Network`?

Reply:
I actually considered moving the whole method to `Network`. First attempt was just to get it to work based on what was implemented in old HNN. Would that work?

Reply:
Sorry I get `Network` and `NetworkBuilder` mixed all the time. I was thinking of `add_tonic_feed` method to both `Network` and `NetworkBuilder`. The previous one adds to `feed_times` after doing sanity checks and the latter instantiates it in Neuron. Wdyt?

Reply:
+1 to the two-pronged approach. Network contains all timing information, Builder is the first time NEURON objects get used. I envision new plotting functions to visualise drives before simulation...

Reply:
@ntolley the reason I thought it could be in `Cell` was because of the Neuron objects involved in the actual implementation of the bias currents. But I agree parameter validation could be in `Network`.

Reply:
okay I decided to keep the parameter validation in `Network` and the rest in `cell.py`. Take a look at what I did